website/hw/hw4.rst

```diff
@@ -180,103 +218,389 @@ As you are stepping through your program, you might want to be able to
+*after* the variable was changed).
 
 When you’re done using gdb, you can exit by typing ``quit`` or by
 pressing Ctrl+D.
 
-Now that you know these commands, let’s have some practice with them!
-Using gdb, answer the following questions on Gradescope:
+If you'd like to practice using these commands a bit more, try doing
+the following (you don't have to answer these questions
+on Gradescope; they are just for you to practice):
+
+- Set a breakpoint at the start of the ``main`` function.
+  Can you find the line(s) where the value of variable ``c`` changes?
+- What is the value pointed to by pointer p after line 42?
+- Set a breakpoint at line 38. Print the value of variable ``a`` and step through lines 38-41,
+  then print the value of variable ``a`` after line 41.
+- What is the return value of the function ``num()`` in hexadecimal?
+
+You can find the answers to these questions at the end of this homework.
+
+The Micro Editor
+----------------
+
+We've seen how GDB works with some simple examples, but now it's time to debug
+a more complex piece of software. We will be using a very simple terminal-based
+editor called ``micro``. This is our version of the ``kilo`` editor, a simple
+but functional text editor that can be implemented in about 1,000 lines of C
+code; if you're interested, you can see a step-by-step guide to writing this
+editor here: https://viewsourcecode.org/snaptoken/kilo/ (please note that you
+do not have to read this to complete this homework; however, if you're interested
+in the inner workings of a text editor, you may find that guide interesting)
+
+``micro`` largely follows the same structure as the ``kilo`` code, except we
+divided it into multiple modules and documented the code following our style guide.
+
+You can build the ``micro`` editor by going into the ``micro/`` directory in
+your repository and running the following::
+
+    cmake -b build/
+    make -C build/
+
+This will generate a ``micro`` executable inside a ``build/`` directory.
+Let's give it a try! Run the following::
+
+    build/micro
+
+This will open the editor with a "blank file". You can start typing to edit
+the file, and you'll notice that you can move around with the arrow keys, use
+the Backspace key, and the Delete key. You can press Ctrl-S to save the file
+(you will be prompted at the bottom of the screen to enter a name; feel free
+to give it any name you want), and then Ctrl-Q to quit.
+
+Let's say we wanted to debug this executable. We actually cannot call GDB like
+before, because the text editor needs to use the entire terminal, so it
+would be impractical to have a GDB prompt interfering with that. So, this will
+allow us to do *remote debugging*, where the program and the debugger are
+run separately (and could even be running in completely different computers,
+and communicating over the network).
+
+To do this, open two terminals. On the first one, run the following from the
```

Comment:
We may want to specify that they need to be ssh'ed into the same linux machine on both  terminal windows